Light Freshness Edit: ASP.NET - Security AUTH (#37122)
dotnet/AspNetCore.Docs: aspnetcore/security/authorization/secure-data.md

:::image type="content" source="secure-data/_static/rick.png" border="false" alt-text="Screenshot showing the user 'rick@contoso.com' signed in to the web app.":::

In the next image, the `manager@contoso.com` user is signed in and has access to the management features:

:::image type="content" source="secure-data/_static/manager1.png" border="false" alt-text="Screenshot showing the user 'manager@contoso.com' signed in to the web app with visibility of management features.":::

A manager can select a contact to see details about the user, as shown in the following image:

:::image type="content" source="secure-data/_static/manager.png" border="false" alt-text="Screenshot shows the manager view of a contact in the web app.":::

The **Approve** and **Reject** options display for managers and administrators only.

> In the following image, the `admin@contoso.com` user is signed and has access to the administration features:

Reviewer: ```suggestion
In the following image, the `admin@contoso.com` user is signed in and has access to the administration features:
```
"Signed" has different meaning in .NET than "Signed in".  In this case it should be "signed in".

Setting the fallback authorization policy to require users to be authenticated protects newly added Razor Pages and controllers. Having authorization required by default is more secure than relying on new controllers and Razor Pages to include the `[Authorize]` attribute.

The <xref:Microsoft.AspNetCore.Authorization.AuthorizationOptions> class also contains the <xref:Microsoft.AspNetCore.Authorization.AuthorizationOptions.DefaultPolicy?displayProperty=nameWithType> property. The `DefaultPolicy` is the policy used with the `[Authorize]` attribute when no policy is specified. `[Authorize]` doesn't contain a named policy, unlike `[Authorize(PolicyName="MyPolicy")]`.

For more information on policies, see <xref:security/authorization/policies>.

As an alternate approach, MVC controllers and Razor Pages can add an authorization filter to require all users be authenticated:

[!code-csharp[](secure-data/samples/final6/Program.cs?name=snippet3&highlight=21-27)]

The preceding code uses an authorization filter, setting the fallback policy uses endpoint routing. Setting the fallback policy is the preferred way to require all users be authenticated.

> Add the [AllowAnonymous](xref:Microsoft.AspNetCore.Authorization.AllowAnonymousAttribute) class to the `Index` and `Privacy` pages so anonymous users can get information about the site before they register:

Reviewer: ```suggestion
Add the [AllowAnonymous](xref:Microsoft.AspNetCore.Authorization.AllowAnonymousAttribute) attribute to the `Index` and `Privacy` pages so anonymous users can get information about the site before they register:
```
AllowAnonymousAttribute is an attribute, not a class in this context of how it's applied here. Calling it a "class" is technically correct at the type level, but misleading in this usage context.

[!code-csharp[](secure-data/samples/final6/Authorization/ContactIsOwnerAuthorizationHandler.cs)]

The `ContactIsOwnerAuthorizationHandler` calls the [context.Succeed](xref:Microsoft.AspNetCore.Authorization.AuthorizationHandlerContext.Succeed%2A) method if the current authenticated user is the contact owner. 

The authorization handler generally:

* Calls the `context.Succeed` method when the requirements are met.

> * When requirements aren't met, it returns the `Task.CompletedTask` status. When the `Task.CompletedTask` status is returned without a prior call to the `context.Success` or `context.Fail` method, the action doesn't indicate a success or failure. Instead, it allows other authorization handlers to run.

Reviewer: ```suggestion
* When requirements aren't met, it returns `Task.CompletedTask`. When `Task.CompletedTask` is returned without a prior call to `context.Succeed` or `context.Fail`, the result isn't a success or failure. Instead, it allows other authorization handlers to run.
```
The original had the error (context.Success instead of context.Succeed), the correct API is context.Succeed. Also, Task.CompletedTask is not a "status", it's a completed Task.

[!code-csharp[](secure-data/samples/final6/Pages/Contacts/Details.cshtml.cs?name=snippet)]

## Add or remove a user to a role
## Add or remove user roles

See [this issue](https://github.com/dotnet/AspNetCore.Docs/issues/8502) for information on:
Updating a user's role assignments lets you control the privileges available to the user. Remove a user from a role and reduce their ability to change data, such as editing or deleting a contact. Add a user to a role and increase their privileges for making global changes. You can also use role assignments to limit user participation, such as muting a user in a chat conversation.

* Removing privileges from a user. For example, muting a user in a chat app.
* Adding privileges to a user.

> <a name="challenge"></a>

Reviewer: Removing this anchor would create a broken link.  See query below. So that topic's link would need to be udpated. (The other two that were removed are OK):

https://github.com/search?q=repo%3Adotnet%2FAspNetCore.Docs+%22secure-data%23challenge%22&type=code

* Registered users can view all of the _Approved_ contact data.
* Registered users can edit/delete their own data.
* Managers can approve/reject contact data. The `Details` view shows **Approve** and **Reject** buttons.
* Administrators can approve/reject and edit/delete all data.

| User                  | Approve/Reject contacts | Options                         |
| --------------------- | :---------------------: | ------------------------------- |
| `test@contoso.com`    | No                      | Edit and delete their data.     |
| `manager@contoso.com` | Yes                     | Edit and delete their data.     |
| `admin@contoso.com`   | Yes                     | Edit and delete ***all*** data. |

Create a contact in the administrator's browser. Copy the URL for delete and edit from the administrator contact. Paste these links into the test user's browser and verify the test user can't perform these operations.

## Create the starter app

* Create a Razor Pages app:

  * Create the app with _Individual Accounts_.
  * Name the app **ContactManager**, so the namespace matches the namespace used in the sample.

>   * Use the `-uld` flag to specify LocalDB instead of SQLite

Reviewer: ```suggestion
  * Use the `-uld` flag to specify LocalDB instead of SQLite.
```
Original was missing a period.
